// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: resources/qualifications/exam.proto

package qualifications

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on QualificationExam with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *QualificationExam) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on QualificationExam with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// QualificationExamMultiError, or nil if none found.
func (m *QualificationExam) ValidateAll() error {
	return m.validate(true)
}

func (m *QualificationExam) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	// no validation rules for QualificationId

	if m.GetQuestions() == nil {
		err := QualificationExamValidationError{
			field:  "Questions",
			reason: "value is required",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if all {
		switch v := interface{}(m.GetQuestions()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, QualificationExamValidationError{
					field:  "Questions",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, QualificationExamValidationError{
					field:  "Questions",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetQuestions()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return QualificationExamValidationError{
				field:  "Questions",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if m.CreatedAt != nil {

		if all {
			switch v := interface{}(m.GetCreatedAt()).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, QualificationExamValidationError{
						field:  "CreatedAt",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, QualificationExamValidationError{
						field:  "CreatedAt",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(m.GetCreatedAt()).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return QualificationExamValidationError{
					field:  "CreatedAt",
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if m.UpdatedAt != nil {

		if all {
			switch v := interface{}(m.GetUpdatedAt()).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, QualificationExamValidationError{
						field:  "UpdatedAt",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, QualificationExamValidationError{
						field:  "UpdatedAt",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(m.GetUpdatedAt()).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return QualificationExamValidationError{
					field:  "UpdatedAt",
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return QualificationExamMultiError(errors)
	}

	return nil
}

// QualificationExamMultiError is an error wrapping multiple validation errors
// returned by QualificationExam.ValidateAll() if the designated constraints
// aren't met.
type QualificationExamMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m QualificationExamMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m QualificationExamMultiError) AllErrors() []error { return m }

// QualificationExamValidationError is the validation error returned by
// QualificationExam.Validate if the designated constraints aren't met.
type QualificationExamValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e QualificationExamValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e QualificationExamValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e QualificationExamValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e QualificationExamValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e QualificationExamValidationError) ErrorName() string {
	return "QualificationExamValidationError"
}

// Error satisfies the builtin error interface
func (e QualificationExamValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sQualificationExam.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = QualificationExamValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = QualificationExamValidationError{}

// Validate checks the field values on ExamSettings with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *ExamSettings) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ExamSettings with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in ExamSettingsMultiError, or
// nil if none found.
func (m *ExamSettings) ValidateAll() error {
	return m.validate(true)
}

func (m *ExamSettings) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if m.GetTime() == nil {
		err := ExamSettingsValidationError{
			field:  "Time",
			reason: "value is required",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if d := m.GetTime(); d != nil {
		dur, err := d.AsDuration(), d.CheckValid()
		if err != nil {
			err = ExamSettingsValidationError{
				field:  "Time",
				reason: "value is not a valid duration",
				cause:  err,
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		} else {

			lt := time.Duration(300*time.Second + 0*time.Nanosecond)
			gte := time.Duration(1036800*time.Second + 0*time.Nanosecond)

			if dur >= lt && dur < gte {
				err := ExamSettingsValidationError{
					field:  "Time",
					reason: "value must be outside range [5m0s, 288h0m0s)",
				}
				if !all {
					return err
				}
				errors = append(errors, err)
			}

		}
	}

	if len(errors) > 0 {
		return ExamSettingsMultiError(errors)
	}

	return nil
}

// ExamSettingsMultiError is an error wrapping multiple validation errors
// returned by ExamSettings.ValidateAll() if the designated constraints aren't met.
type ExamSettingsMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ExamSettingsMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ExamSettingsMultiError) AllErrors() []error { return m }

// ExamSettingsValidationError is the validation error returned by
// ExamSettings.Validate if the designated constraints aren't met.
type ExamSettingsValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ExamSettingsValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ExamSettingsValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ExamSettingsValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ExamSettingsValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ExamSettingsValidationError) ErrorName() string { return "ExamSettingsValidationError" }

// Error satisfies the builtin error interface
func (e ExamSettingsValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sExamSettings.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ExamSettingsValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ExamSettingsValidationError{}

// Validate checks the field values on ExamQuestions with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *ExamQuestions) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ExamQuestions with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in ExamQuestionsMultiError, or
// nil if none found.
func (m *ExamQuestions) ValidateAll() error {
	return m.validate(true)
}

func (m *ExamQuestions) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	for idx, item := range m.GetData() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, ExamQuestionsValidationError{
						field:  fmt.Sprintf("Data[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, ExamQuestionsValidationError{
						field:  fmt.Sprintf("Data[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ExamQuestionsValidationError{
					field:  fmt.Sprintf("Data[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return ExamQuestionsMultiError(errors)
	}

	return nil
}

// ExamQuestionsMultiError is an error wrapping multiple validation errors
// returned by ExamQuestions.ValidateAll() if the designated constraints
// aren't met.
type ExamQuestionsMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ExamQuestionsMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ExamQuestionsMultiError) AllErrors() []error { return m }

// ExamQuestionsValidationError is the validation error returned by
// ExamQuestions.Validate if the designated constraints aren't met.
type ExamQuestionsValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ExamQuestionsValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ExamQuestionsValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ExamQuestionsValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ExamQuestionsValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ExamQuestionsValidationError) ErrorName() string { return "ExamQuestionsValidationError" }

// Error satisfies the builtin error interface
func (e ExamQuestionsValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sExamQuestions.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ExamQuestionsValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ExamQuestionsValidationError{}

// Validate checks the field values on ExamQuestion with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *ExamQuestion) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ExamQuestion with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in ExamQuestionMultiError, or
// nil if none found.
func (m *ExamQuestion) ValidateAll() error {
	return m.validate(true)
}

func (m *ExamQuestion) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	switch v := m.Data.(type) {
	case *ExamQuestion_Separator:
		if v == nil {
			err := ExamQuestionValidationError{
				field:  "Data",
				reason: "oneof value cannot be a typed-nil",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}
		// no validation rules for Separator
	case *ExamQuestion_Yesno:
		if v == nil {
			err := ExamQuestionValidationError{
				field:  "Data",
				reason: "oneof value cannot be a typed-nil",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}

		if all {
			switch v := interface{}(m.GetYesno()).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, ExamQuestionValidationError{
						field:  "Yesno",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, ExamQuestionValidationError{
						field:  "Yesno",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(m.GetYesno()).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ExamQuestionValidationError{
					field:  "Yesno",
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	case *ExamQuestion_FreeText:
		if v == nil {
			err := ExamQuestionValidationError{
				field:  "Data",
				reason: "oneof value cannot be a typed-nil",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}

		if all {
			switch v := interface{}(m.GetFreeText()).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, ExamQuestionValidationError{
						field:  "FreeText",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, ExamQuestionValidationError{
						field:  "FreeText",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(m.GetFreeText()).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ExamQuestionValidationError{
					field:  "FreeText",
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	case *ExamQuestion_MultipleChoice:
		if v == nil {
			err := ExamQuestionValidationError{
				field:  "Data",
				reason: "oneof value cannot be a typed-nil",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}

		if all {
			switch v := interface{}(m.GetMultipleChoice()).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, ExamQuestionValidationError{
						field:  "MultipleChoice",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, ExamQuestionValidationError{
						field:  "MultipleChoice",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(m.GetMultipleChoice()).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ExamQuestionValidationError{
					field:  "MultipleChoice",
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	default:
		_ = v // ensures v is used
	}

	if len(errors) > 0 {
		return ExamQuestionMultiError(errors)
	}

	return nil
}

// ExamQuestionMultiError is an error wrapping multiple validation errors
// returned by ExamQuestion.ValidateAll() if the designated constraints aren't met.
type ExamQuestionMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ExamQuestionMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ExamQuestionMultiError) AllErrors() []error { return m }

// ExamQuestionValidationError is the validation error returned by
// ExamQuestion.Validate if the designated constraints aren't met.
type ExamQuestionValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ExamQuestionValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ExamQuestionValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ExamQuestionValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ExamQuestionValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ExamQuestionValidationError) ErrorName() string { return "ExamQuestionValidationError" }

// Error satisfies the builtin error interface
func (e ExamQuestionValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sExamQuestion.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ExamQuestionValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ExamQuestionValidationError{}

// Validate checks the field values on ExamQuestionYesNo with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *ExamQuestionYesNo) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ExamQuestionYesNo with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ExamQuestionYesNoMultiError, or nil if none found.
func (m *ExamQuestionYesNo) ValidateAll() error {
	return m.validate(true)
}

func (m *ExamQuestionYesNo) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return ExamQuestionYesNoMultiError(errors)
	}

	return nil
}

// ExamQuestionYesNoMultiError is an error wrapping multiple validation errors
// returned by ExamQuestionYesNo.ValidateAll() if the designated constraints
// aren't met.
type ExamQuestionYesNoMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ExamQuestionYesNoMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ExamQuestionYesNoMultiError) AllErrors() []error { return m }

// ExamQuestionYesNoValidationError is the validation error returned by
// ExamQuestionYesNo.Validate if the designated constraints aren't met.
type ExamQuestionYesNoValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ExamQuestionYesNoValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ExamQuestionYesNoValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ExamQuestionYesNoValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ExamQuestionYesNoValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ExamQuestionYesNoValidationError) ErrorName() string {
	return "ExamQuestionYesNoValidationError"
}

// Error satisfies the builtin error interface
func (e ExamQuestionYesNoValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sExamQuestionYesNo.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ExamQuestionYesNoValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ExamQuestionYesNoValidationError{}

// Validate checks the field values on ExamQuestionText with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *ExamQuestionText) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ExamQuestionText with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ExamQuestionTextMultiError, or nil if none found.
func (m *ExamQuestionText) ValidateAll() error {
	return m.validate(true)
}

func (m *ExamQuestionText) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for MinLength

	// no validation rules for MaxLength

	if len(errors) > 0 {
		return ExamQuestionTextMultiError(errors)
	}

	return nil
}

// ExamQuestionTextMultiError is an error wrapping multiple validation errors
// returned by ExamQuestionText.ValidateAll() if the designated constraints
// aren't met.
type ExamQuestionTextMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ExamQuestionTextMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ExamQuestionTextMultiError) AllErrors() []error { return m }

// ExamQuestionTextValidationError is the validation error returned by
// ExamQuestionText.Validate if the designated constraints aren't met.
type ExamQuestionTextValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ExamQuestionTextValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ExamQuestionTextValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ExamQuestionTextValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ExamQuestionTextValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ExamQuestionTextValidationError) ErrorName() string { return "ExamQuestionTextValidationError" }

// Error satisfies the builtin error interface
func (e ExamQuestionTextValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sExamQuestionText.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ExamQuestionTextValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ExamQuestionTextValidationError{}

// Validate checks the field values on ExamQuestionChoices with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ExamQuestionChoices) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ExamQuestionChoices with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ExamQuestionChoicesMultiError, or nil if none found.
func (m *ExamQuestionChoices) ValidateAll() error {
	return m.validate(true)
}

func (m *ExamQuestionChoices) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Multi

	if len(m.GetChoices()) > 10 {
		err := ExamQuestionChoicesValidationError{
			field:  "Choices",
			reason: "value must contain no more than 10 item(s)",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if m.Limit != nil {
		// no validation rules for Limit
	}

	if len(errors) > 0 {
		return ExamQuestionChoicesMultiError(errors)
	}

	return nil
}

// ExamQuestionChoicesMultiError is an error wrapping multiple validation
// errors returned by ExamQuestionChoices.ValidateAll() if the designated
// constraints aren't met.
type ExamQuestionChoicesMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ExamQuestionChoicesMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ExamQuestionChoicesMultiError) AllErrors() []error { return m }

// ExamQuestionChoicesValidationError is the validation error returned by
// ExamQuestionChoices.Validate if the designated constraints aren't met.
type ExamQuestionChoicesValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ExamQuestionChoicesValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ExamQuestionChoicesValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ExamQuestionChoicesValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ExamQuestionChoicesValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ExamQuestionChoicesValidationError) ErrorName() string {
	return "ExamQuestionChoicesValidationError"
}

// Error satisfies the builtin error interface
func (e ExamQuestionChoicesValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sExamQuestionChoices.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ExamQuestionChoicesValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ExamQuestionChoicesValidationError{}

// Validate checks the field values on QualificationExamResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *QualificationExamResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on QualificationExamResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// QualificationExamResponseMultiError, or nil if none found.
func (m *QualificationExamResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *QualificationExamResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	// no validation rules for QualificationId

	if all {
		switch v := interface{}(m.GetResponses()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, QualificationExamResponseValidationError{
					field:  "Responses",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, QualificationExamResponseValidationError{
					field:  "Responses",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetResponses()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return QualificationExamResponseValidationError{
				field:  "Responses",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if m.StartedAt != nil {

		if all {
			switch v := interface{}(m.GetStartedAt()).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, QualificationExamResponseValidationError{
						field:  "StartedAt",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, QualificationExamResponseValidationError{
						field:  "StartedAt",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(m.GetStartedAt()).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return QualificationExamResponseValidationError{
					field:  "StartedAt",
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if m.EndedAt != nil {

		if all {
			switch v := interface{}(m.GetEndedAt()).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, QualificationExamResponseValidationError{
						field:  "EndedAt",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, QualificationExamResponseValidationError{
						field:  "EndedAt",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(m.GetEndedAt()).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return QualificationExamResponseValidationError{
					field:  "EndedAt",
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return QualificationExamResponseMultiError(errors)
	}

	return nil
}

// QualificationExamResponseMultiError is an error wrapping multiple validation
// errors returned by QualificationExamResponse.ValidateAll() if the
// designated constraints aren't met.
type QualificationExamResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m QualificationExamResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m QualificationExamResponseMultiError) AllErrors() []error { return m }

// QualificationExamResponseValidationError is the validation error returned by
// QualificationExamResponse.Validate if the designated constraints aren't met.
type QualificationExamResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e QualificationExamResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e QualificationExamResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e QualificationExamResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e QualificationExamResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e QualificationExamResponseValidationError) ErrorName() string {
	return "QualificationExamResponseValidationError"
}

// Error satisfies the builtin error interface
func (e QualificationExamResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sQualificationExamResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = QualificationExamResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = QualificationExamResponseValidationError{}

// Validate checks the field values on ExamResponses with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *ExamResponses) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ExamResponses with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in ExamResponsesMultiError, or
// nil if none found.
func (m *ExamResponses) ValidateAll() error {
	return m.validate(true)
}

func (m *ExamResponses) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	for idx, item := range m.GetData() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, ExamResponsesValidationError{
						field:  fmt.Sprintf("Data[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, ExamResponsesValidationError{
						field:  fmt.Sprintf("Data[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ExamResponsesValidationError{
					field:  fmt.Sprintf("Data[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return ExamResponsesMultiError(errors)
	}

	return nil
}

// ExamResponsesMultiError is an error wrapping multiple validation errors
// returned by ExamResponses.ValidateAll() if the designated constraints
// aren't met.
type ExamResponsesMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ExamResponsesMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ExamResponsesMultiError) AllErrors() []error { return m }

// ExamResponsesValidationError is the validation error returned by
// ExamResponses.Validate if the designated constraints aren't met.
type ExamResponsesValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ExamResponsesValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ExamResponsesValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ExamResponsesValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ExamResponsesValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ExamResponsesValidationError) ErrorName() string { return "ExamResponsesValidationError" }

// Error satisfies the builtin error interface
func (e ExamResponsesValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sExamResponses.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ExamResponsesValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ExamResponsesValidationError{}

// Validate checks the field values on ExamResponse with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *ExamResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ExamResponse with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in ExamResponseMultiError, or
// nil if none found.
func (m *ExamResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ExamResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	switch v := m.Data.(type) {
	case *ExamResponse_Separator:
		if v == nil {
			err := ExamResponseValidationError{
				field:  "Data",
				reason: "oneof value cannot be a typed-nil",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}
		// no validation rules for Separator
	case *ExamResponse_Yesno:
		if v == nil {
			err := ExamResponseValidationError{
				field:  "Data",
				reason: "oneof value cannot be a typed-nil",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}

		if all {
			switch v := interface{}(m.GetYesno()).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, ExamResponseValidationError{
						field:  "Yesno",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, ExamResponseValidationError{
						field:  "Yesno",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(m.GetYesno()).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ExamResponseValidationError{
					field:  "Yesno",
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	case *ExamResponse_FreeText:
		if v == nil {
			err := ExamResponseValidationError{
				field:  "Data",
				reason: "oneof value cannot be a typed-nil",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}

		if all {
			switch v := interface{}(m.GetFreeText()).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, ExamResponseValidationError{
						field:  "FreeText",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, ExamResponseValidationError{
						field:  "FreeText",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(m.GetFreeText()).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ExamResponseValidationError{
					field:  "FreeText",
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	case *ExamResponse_MultipleChoice:
		if v == nil {
			err := ExamResponseValidationError{
				field:  "Data",
				reason: "oneof value cannot be a typed-nil",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}

		if all {
			switch v := interface{}(m.GetMultipleChoice()).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, ExamResponseValidationError{
						field:  "MultipleChoice",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, ExamResponseValidationError{
						field:  "MultipleChoice",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(m.GetMultipleChoice()).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ExamResponseValidationError{
					field:  "MultipleChoice",
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	default:
		_ = v // ensures v is used
	}

	if len(errors) > 0 {
		return ExamResponseMultiError(errors)
	}

	return nil
}

// ExamResponseMultiError is an error wrapping multiple validation errors
// returned by ExamResponse.ValidateAll() if the designated constraints aren't met.
type ExamResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ExamResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ExamResponseMultiError) AllErrors() []error { return m }

// ExamResponseValidationError is the validation error returned by
// ExamResponse.Validate if the designated constraints aren't met.
type ExamResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ExamResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ExamResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ExamResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ExamResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ExamResponseValidationError) ErrorName() string { return "ExamResponseValidationError" }

// Error satisfies the builtin error interface
func (e ExamResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sExamResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ExamResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ExamResponseValidationError{}

// Validate checks the field values on ExamResponseYesNo with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *ExamResponseYesNo) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ExamResponseYesNo with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ExamResponseYesNoMultiError, or nil if none found.
func (m *ExamResponseYesNo) ValidateAll() error {
	return m.validate(true)
}

func (m *ExamResponseYesNo) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Value

	if len(errors) > 0 {
		return ExamResponseYesNoMultiError(errors)
	}

	return nil
}

// ExamResponseYesNoMultiError is an error wrapping multiple validation errors
// returned by ExamResponseYesNo.ValidateAll() if the designated constraints
// aren't met.
type ExamResponseYesNoMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ExamResponseYesNoMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ExamResponseYesNoMultiError) AllErrors() []error { return m }

// ExamResponseYesNoValidationError is the validation error returned by
// ExamResponseYesNo.Validate if the designated constraints aren't met.
type ExamResponseYesNoValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ExamResponseYesNoValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ExamResponseYesNoValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ExamResponseYesNoValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ExamResponseYesNoValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ExamResponseYesNoValidationError) ErrorName() string {
	return "ExamResponseYesNoValidationError"
}

// Error satisfies the builtin error interface
func (e ExamResponseYesNoValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sExamResponseYesNo.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ExamResponseYesNoValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ExamResponseYesNoValidationError{}

// Validate checks the field values on ExamResponseText with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *ExamResponseText) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ExamResponseText with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ExamResponseTextMultiError, or nil if none found.
func (m *ExamResponseText) ValidateAll() error {
	return m.validate(true)
}

func (m *ExamResponseText) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Text

	if len(errors) > 0 {
		return ExamResponseTextMultiError(errors)
	}

	return nil
}

// ExamResponseTextMultiError is an error wrapping multiple validation errors
// returned by ExamResponseText.ValidateAll() if the designated constraints
// aren't met.
type ExamResponseTextMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ExamResponseTextMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ExamResponseTextMultiError) AllErrors() []error { return m }

// ExamResponseTextValidationError is the validation error returned by
// ExamResponseText.Validate if the designated constraints aren't met.
type ExamResponseTextValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ExamResponseTextValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ExamResponseTextValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ExamResponseTextValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ExamResponseTextValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ExamResponseTextValidationError) ErrorName() string { return "ExamResponseTextValidationError" }

// Error satisfies the builtin error interface
func (e ExamResponseTextValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sExamResponseText.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ExamResponseTextValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ExamResponseTextValidationError{}

// Validate checks the field values on ExamResponseChoices with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ExamResponseChoices) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ExamResponseChoices with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ExamResponseChoicesMultiError, or nil if none found.
func (m *ExamResponseChoices) ValidateAll() error {
	return m.validate(true)
}

func (m *ExamResponseChoices) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return ExamResponseChoicesMultiError(errors)
	}

	return nil
}

// ExamResponseChoicesMultiError is an error wrapping multiple validation
// errors returned by ExamResponseChoices.ValidateAll() if the designated
// constraints aren't met.
type ExamResponseChoicesMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ExamResponseChoicesMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ExamResponseChoicesMultiError) AllErrors() []error { return m }

// ExamResponseChoicesValidationError is the validation error returned by
// ExamResponseChoices.Validate if the designated constraints aren't met.
type ExamResponseChoicesValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ExamResponseChoicesValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ExamResponseChoicesValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ExamResponseChoicesValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ExamResponseChoicesValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ExamResponseChoicesValidationError) ErrorName() string {
	return "ExamResponseChoicesValidationError"
}

// Error satisfies the builtin error interface
func (e ExamResponseChoicesValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sExamResponseChoices.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ExamResponseChoicesValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ExamResponseChoicesValidationError{}
